ibis/sql/postgres/udf/api.py

```diff
@@ -0,0 +1,218 @@
+    Decimal: 'NUMERIC',
+    bytes: 'BYTEA',
+    str: sql_default_type,
+}
+
+pytype_ibistype = {
+    bool: datatypes.Boolean(),
+    int: datatypes.Int32(),
+    float: datatypes.Float(),
+    Decimal: datatypes.Float(),
+    bytes: datatypes.Binary(),
+    str: datatypes.String(),
+}
+
+ibistype_pytype = {v: k for k, v in pytype_ibistype.items()}
+
+
+def get_sqltype(type_):
+    """Map input to the string specifying the Postgres data type
+    (as is used in SQL defining UDF signatures)
+
+    :param type_: str, a Python data type,
+                  or an ibis DataType (instance or class)
+    :return: string symbol of Postgres data type
+    """
+    if isinstance(type_, str) and type_ in pytype_sql.values():
+        return type_
+    elif type_ in pytype_sql.keys():
+        return pytype_sql[type_]
+    elif type_ in ibistype_pytype:
+        return pytype_sql[ibistype_pytype[type_]]
+    elif type_ in set(map(type, ibistype_pytype.keys())):
+        return pytype_sql[ibistype_pytype[type_()]]
+    else:
+        raise ValueError(
+            "Postgres data type not defined for: {}".format(type_)
+        )
+
+
+def create_udf_node(name, fields):
+    """Create a new UDF node type.
+
+    Parameters
+    ----------
+    name : str
+        Then name of the UDF node
+    fields : OrderedDict
+        Mapping of class member name to definition
+
+    Returns
+    -------
+    result : type
+        A new PostgresUDFNode subclass
+    """
+    definition = next(_udf_name_cache[name])
+    external_name = '{}_{:d}'.format(name, definition)
+    return type(external_name, (PostgresUDFNode,), fields)
+
+
+def existing_udf(name,
+                 input_types,
+                 output_type,
+                 schema=None,
+                 parameter_names=None):
+    """Create a ibis function that refers to an existing Postgres UDF already
+    defined in database
+
+    Parameters
+    ----------
+    name: str
+    input_types : List[DataType]
+    output_type : DataType
+    schema: str - optionally specify the schema that the UDF is defined in
+    parameter_names: str - give names to the arguments of the UDF
+
+    Returns
+    -------
+    wrapper : Callable
+        The wrapped function
+    """
+    if parameter_names is None:
+        parameter_names = ['v{}'.format(i) for i in range(len(input_types))]
+    else:
+        assert len(input_types) == len(parameter_names)
+
+    udf_node_fields = collections.OrderedDict([
+        (name, Arg(rlz.value(type_)))
+        for name, type_ in zip(parameter_names, input_types)
+    ] + [
+        (
+            'output_type',
+            lambda self, output_type=output_type: rlz.shape_like(
+                self.args, dtype=output_type
+            )
+        )
+    ])
+    udf_node_fields['resolve_name'] = lambda self: name
+
+    udf_node = create_udf_node(name, udf_node_fields)
+
+    def _translate_udf(t, expr):
+        func_obj = sqlalchemy.func
+        if schema is not None:
```

Comment:
Can we set `schema` to some kind of valid default (I think it's actually `'default'` in postgres) so that we don't need to first check if `schema` is None? You could then write this as `func_obj = getattr(getattr(sa.func, schema), name)`.

Reply:
I'd like to keep the possibility open for not qualifying the schema. While there usually is a default schema called `public` in Postgres, there is also a way (in postgres and in `sqlalchemy`) of setting the schema search path so that another non-`public` schema is used when unqualified.